source/dem/ray_tracing.cc

```diff
@@ -0,0 +1,768 @@
+  std::vector<Point<3>> insertion_points_on_proc;
+
+  // For each photon, we store the initial location of the photon as a particle
+  // property. This initial location is used to identify the right intersection
+  // point when multiple intersections are found. The logic used therein is that
+  // the intersection point which is closest to the initial location is the
+  // correct one. We also store the displacement direction unit vector in this
+  // vector since each photon has its own slightly off set from the
+  std::vector<std::vector<double>> photon_properties;
+
+  // Create variables for readability
+  const std::vector<unsigned int> &n_photons_each_directions =
+    parameters.ray_tracing_info.n_photons_each_directions;
+
+  // Starting point
+  const Point<dim> starting_insertion_point =
+    parameters.ray_tracing_info.starting_point;
+
+  // Insertion directions
+  const Tensor<1, dim> first_dir =
+    parameters.ray_tracing_info.insertion_directions_units_vector.at(0);
+  const Tensor<1, dim> second_dir =
+    parameters.ray_tracing_info.insertion_directions_units_vector.at(1);
+  const Tensor<1, dim> third_dir =
+    parameters.ray_tracing_info.insertion_directions_units_vector.at(2);
+
+  // Steps
+  const double step_first_dir =
+    parameters.ray_tracing_info.step_between_photons_each_directions.at(0);
+  const double step_second_dir =
+    parameters.ray_tracing_info.step_between_photons_each_directions.at(1);
+  const double step_third_dir =
+    parameters.ray_tracing_info.step_between_photons_each_directions.at(2);
+
+  // Reference displacement direction
+  const Tensor<1, 3> ref_displacement_dir =
+    parameters.ray_tracing_info.ref_displacement_tensor_unit;
+
+  // Total number of photons to insert
+  const unsigned int n_total_photons_to_insert =
+    n_photons_each_directions.at(0) * n_photons_each_directions.at(1) *
+    n_photons_each_directions.at(2);
+
+  // Distributing photons between processors
+  const unsigned int base_photons_per_proc =
+    n_total_photons_to_insert / n_mpi_processes;
+  const unsigned int remainder = n_total_photons_to_insert % n_mpi_processes;
+
+  const unsigned int n_photons_to_insert_this_proc =
+    base_photons_per_proc + (this_mpi_process < remainder ? 1 : 0);
+
+  // First and last photon id on this processor. This will be used to find the
+  // position of the photons.
+  const unsigned int first_id =
+    this_mpi_process * base_photons_per_proc +
+    (this_mpi_process < remainder ? this_mpi_process : remainder);
+  const unsigned int last_id = first_id + n_photons_to_insert_this_proc;
+
+  // Generate the random offset in the insertion position and the
+  // displacement direction
+  std::vector<double> random_number_angular_1; // From 0 to 2*pi
+  std::vector<double> random_number_angular_2; // From 0 to max_angular_offset
+  std::vector<double> random_number_position;  // From 0 to max_insertion_offset
```

Comment:
why is the random number generated for angular? Could you provide a bit more explanation of what is being done here.